File: cpp/454_4SumII.cpp
class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        unordered_map<int, int> sums;
        for (int a : A) {
            for (int b : B) {
                sums[a + b] = sums.find(a + b) == sums.end() ? 1 : sums[a + b] + 1;
            }
        }
        
        int count = 0;
        for (int c : C) {
            for (int d : D) {
                if (sums.find(-(c+d)) != sums.end())
                    count += sums[-(c+d)];
            }
        }
        return count;
    }
};

File: python/454_4SumII.py
class Solution:
    def fourSumCount(self, A: List[int], B: List[int], C: List[int], D: List[int]) -> int:
        sums = {}
        for a in A:
            for b in B:
                sums[a + b] = sums[a + b] + 1 if a + b in sums else 1
        
        count = 0
        for c in C:
            for d in D:
                if -(c + d) in sums: count += sums[-(c + d)]
        return count
